File: Z_Testing.py
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <cuda_runtime.h>



__global__ void matrixMultiplication(int* A, int* B, int* C, int n) {
    int row = blockIdx.y * blockDim.y + threadIdx.y;
    int col = blockIdx.x * blockDim.x + threadIdx.x;

    if (row < n && col < n) {
        int sum = 0;
        for (int k = 0; k < n; ++k) {
            sum += A[row * n + k] * B[k * n + col];
        }
        C[row * n + col] = sum;
    }
}




int main() {
    int n;
    std::cin >> n;

    int* h_A = new int[n * n];
    int* h_B = new int[n * n];
    int* h_C = new int[n * n];

    // Initialize matrices h_A and h_B with random values
    srand(time(NULL));
    for (int i = 0; i < n * n; ++i) {
        h_A[i] = rand() % 1000 + 1;
        h_B[i] = rand() % 1000 + 1;
    }

    int* d_A;
    int* d_B;
    int* d_C;

    // Allocate memory on the GPU
    cudaMalloc((void**)&d_A, sizeof(int) * n * n);
    cudaMalloc((void**)&d_B, sizeof(int) * n * n);
    cudaMalloc((void**)&d_C, sizeof(int) * n * n);

    // Copy data from host to device
    cudaMemcpy(d_A, h_A, sizeof(int) * n * n, cudaMemcpyHostToDevice);
    cudaMemcpy(d_B, h_B, sizeof(int) * n * n, cudaMemcpyHostToDevice);

    // Define thread and block dimensions
    dim3 threadsPerBlock(16, 16);
    dim3 numBlocks((n + 15) / 16, (n + 15) / 16);

    // Launch the matrix multiplication kernel
    matrixMultiplication<<<numBlocks, threadsPerBlock>>>(d_A, d_B, d_C, n);

    // Copy the result from device to host
    cudaMemcpy(h_C, d_C, sizeof(int) * n * n, cudaMemcpyDeviceToHost);

    // Print the result matrix h_C if needed

    // Free GPU memory
    cudaFree(d_A);
    cudaFree(d_B);
    cudaFree(d_C);

    // Free CPU memory
    delete[] h_A;
    delete[] h_B;
    delete[] h_C;

    return 0;
}

File: Z_Testing.java
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <cuda_runtime.h>

#define ll long long
#define fr(i,n) for(int i=0;i<n;i++)

__global__ void matrixMul(int *mat1, int *mat2, int *result, int n) {
    int row = blockIdx.y * blockDim.y + threadIdx.y;
    int col = blockIdx.x * blockDim.x + threadIdx.x;

    if (row < n && col < n) {
        int val = 0;
        for (int k = 0; k < n; ++k) {
            val += mat1[row * n + k] * mat2[k * n + col];
        }
        result[row * n + col] = val;
    }
}

int main() {
    int n;
    std::cin >> n;

    int *mat1, *mat2, *result;
    int size = n * n * sizeof(int);

    // Allocate memory on the GPU
    cudaMalloc((void**)&mat1, size);
    cudaMalloc((void**)&mat2, size);
    cudaMalloc((void**)&result, size);

    // Generate random values on the CPU
    srand(time(NULL));
    std::vector<int> host_mat1(n * n);
    std::vector<int> host_mat2(n * n);

    for (int i = 0; i < n * n; ++i) {
        host_mat1[i] = rand() % 1000 + 1;
        host_mat2[i] = rand() % 1000 + 1;
    }

    // Copy data from the CPU to the GPU
    cudaMemcpy(mat1, host_mat1.data(), size, cudaMemcpyHostToDevice);
    cudaMemcpy(mat2, host_mat2.data(), size, cudaMemcpyHostToDevice);

    // Define block and grid dimensions
    dim3 blockSize(16, 16);
    dim3 gridSize((n + blockSize.x - 1) / blockSize.x, (n + blockSize.y - 1) / blockSize.y);

    // Launch the CUDA kernel
    matrixMul<<<gridSize, blockSize>>>(mat1, mat2, result, n);

    // Copy the result back to the CPU
    std::vector<int> host_result(n * n);
    cudaMemcpy(host_result.data(), result, size, cudaMemcpyDeviceToHost);

    // Print the result matrix (if needed)
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cout << host_result[i * n + j] << " ";
        }
        std::cout << std::endl;
    }

    // Free GPU memory
    cudaFree(mat1);
    cudaFree(mat2);
    cudaFree(result);

    return 0;
}

















#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <cuda_runtime.h>

using namespace std; // Add this line to use the std namespace

#define ll long long
#define fr(i,n) for(int i=0;i<n;i++)

__global__ void matrixMul(int *mat1, int *mat2, int *result, int n) {
    // ... (rest of the code remains the same)
}

int main() {
    // ... (rest of the code remains the same)
    
    // Print the result matrix (if needed)
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cout << host_result[i * n + j] << " "; // Now you can use cout without std:: prefix
        }
        cout << endl;
    }

    // ... (rest of the code remains the same)

    return 0;
}























#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <cuda_runtime.h>

using namespace std; // Add this line to use the std namespace

#define ll long long
#define fr(i,n) for(int i=0;i<n;i++)

__global__ void matrixMul(int *mat1, int *mat2, int *result, int n) {
    int row = blockIdx.y * blockDim.y + threadIdx.y;
    int col = blockIdx.x * blockDim.x + threadIdx.x;

    if (row < n && col < n) {
        int val = 0;
        for (int k = 0; k < n; ++k) {
            val += mat1[row * n + k] * mat2[k * n + col];
        }
        result[row * n + col] = val;
    }
}

int main() {
    int n;
    cin >> n;

    int *mat1, *mat2, *result;
    int size = n * n * sizeof(int);

    // Allocate memory on the GPU
    cudaMalloc((void**)&mat1, size);
    cudaMalloc((void**)&mat2, size);
    cudaMalloc((void**)&result, size);

    // Generate random values on the CPU
    srand(time(NULL));
    vector<int> host_mat1(n * n);
    vector<int> host_mat2(n * n);

    for (int i = 0; i < n * n; ++i) {
        host_mat1[i] = rand() % 1000 + 1;
        host_mat2[i] = rand() % 1000 + 1;
    }

    // Copy data from the CPU to the GPU
    cudaMemcpy(mat1, host_mat1.data(), size, cudaMemcpyHostToDevice);
    cudaMemcpy(mat2, host_mat2.data(), size, cudaMemcpyHostToDevice);

    // Define block and grid dimensions
    dim3 blockSize(16, 16);
    dim3 gridSize((n + blockSize.x - 1) / blockSize.x, (n + blockSize.y - 1) / blockSize.y);

    // Launch the CUDA kernel
    matrixMul<<<gridSize, blockSize>>>(mat1, mat2, result, n);

    // Copy the result back to the CPU
    vector<int> host_result(n * n);
    cudaMemcpy(host_result.data(), result, size, cudaMemcpyDeviceToHost);

    // Print the result matrix (if needed)
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cout << host_result[i * n + j] << " ";
        }
        cout << endl;
    }

    // Free GPU memory
    cudaFree(mat1);
    cudaFree(mat2);
    cudaFree(result);

    return 0;
}
